src/rules/no_invalid_triple_slash_reference.rs

```diff
@@ -0,0 +1,223 @@
+// Copyright 2020-2021 the Deno authors. All rights reserved. MIT license.
+use super::{Context, LintRule, ProgramRef};
+use derive_more::Display;
+use once_cell::sync::Lazy;
+use regex::Regex;
+use swc_common::comments::{Comment, CommentKind};
+use swc_common::Span;
+
+pub struct NoInvalidTripleSlashReference;
+
+const CODE: &str = "no-invalid-triple-slash-reference";
+
+#[derive(Display)]
+enum NoInvalidTripleSlashReferenceMessage {
+  #[display(
+    fmt = "triple-slash references with `path` have no effect in Deno"
+  )]
+  Path,
+}
+
+#[derive(Display)]
+enum NoInvalidTripleSlashReferenceHint {
+  #[display(
+    fmt = r#"Replace `path` with `types`, like `/// <reference types="{}" />`"#,
+    _0
+  )]
+  Replace(String),
+}
+
+impl LintRule for NoInvalidTripleSlashReference {
+  fn new() -> Box<Self> {
+    Box::new(NoInvalidTripleSlashReference)
+  }
+
+  fn tags(&self) -> &'static [&'static str] {
+    &["recommended"]
+  }
+
+  fn code(&self) -> &'static str {
+    CODE
+  }
+
+  fn lint_program(&self, _context: &mut Context, _program: ProgramRef<'_>) {
+    unreachable!();
+  }
+
+  fn lint_program_with_ast_view(
+    &self,
+    context: &mut Context,
+    _program: dprint_swc_ecma_ast_view::Program<'_>,
+  ) {
+    for (span, path) in context.all_comments().filter_map(check_comment) {
+      context.add_diagnostic_with_hint(
+        span,
+        CODE,
+        NoInvalidTripleSlashReferenceMessage::Path,
+        NoInvalidTripleSlashReferenceHint::Replace(path),
+      );
+    }
+  }
+
+  fn docs(&self) -> &'static str {
+    r#"Warns the usage of triple-slash references directive of `path`.
+
+Deno supports the triple-slash reference `types` directive, which is useful for
+telling the TypeScript compiler the location of a type definition file that
+corresponds to a certain JavaScript file.
+
+However, in the Deno manual of the previous versions (e.g. [v1.9.2]), there was
+a wrong statement describing that one should use the `path` directive. Actually,
+the `types` directive should be used. See [the latest manual] for more detail.
+
+[v1.9.2]: https://deno.land/manual@v1.9.2/typescript/types#using-the-triple-slash-reference-directive
+[the latest manual]: https://deno.land/manual/typescript/types#using-the-triple-slash-reference-directive
+
+This lint rule detects such wrong usage of the `path` directive and suggests
+replacing it with the `types` directive.
+
+### Invalid:
+```javascript
+/// <reference path="./mod.d.ts" />
+
+// ... the rest of the JavaScript ...
+```
+
+### Valid:
+```javascript
+/// <reference types="./mod.d.ts" />
+
+// ... the rest of the JavaScript ...
+```
+"#
+  }
+}
+
+/// Returns `Some` if the comment should be reported.
+fn check_comment(comment: &Comment) -> Option<(Span, String)> {
+  if comment.kind == CommentKind::Block {
+    return None;
+  }
+
+  static INVALID_TSR_REGEX: Lazy<Regex> = Lazy::new(|| {
+    Regex::new(r#"^/\s*<reference\s*path\s*=\s*["|'](?P<path>.*)["|']\s*/>"#)
+      .unwrap()
```

Comment:
Rather than blocking only "reference path", could it be an allow-list (presumably we want to allow only "reference types=" and "reference lib=") so that things like 
```
/// <reference foo="bar">
```
would also trigger this rule?

Reply:
I agree with @ry, we only want to allow `/// <reference types="..." />` in JavaScript/JSX files (we shouldn't have the diagnostic for TypeScript/TSX files).

Also we should be consistent with the logic that is in the `module_graph`:

https://github.com/denoland/deno/blob/59237d195f6afa0d8927128751ba5ff4562d7cce/cli/module_graph.rs#L204-L224

Reply:
allow-list sounds good.
And I've noticed that my wording in the doc, "triple-slash references with `path` have no effect in Deno", is likely wrong... actually the `path` directive allows adding a dependency to a module in Deno as well.

So my current understanding is that this lint rule should be triggered in cases such as:

- triple-slash references in which none of the allowed directives (`path`, `types`, `lib`) is specified (i.e. invalid format)
- triple-slash references with `path` iff the file is JS or JSX (i.e. the `path` directive is used unintentionally)

I'm fixing it right now.

Reply:
To be clear, in TypeScript files (TypeScript/TSX) triple-slash directives are allowed and work.

In JavaScript files (JSX, TSX), only the `types` directive is allowed.

Reply:
Ah OK, I misunderstood the second point. Thanks for clarifying :)

Reply:
@ry @kitsonk 
Fixed. Now, when the target file is JS-like, this rule tries to detect the directives other than `types`, and when it's not JS-like, it checks if there is an invalid format of triple-slash reference.
Please take a look again 😃